toolchain/check/impl_validation.cpp

```diff
@@ -244,90 +239,104 @@ static auto DiagnoseFinalImplsOverlapOutsideMatchFirst(Context& context,
-      auto impl_b = GetImplInfo(context, impl_b_id);
 
       if (!impl_a.is_final && !impl_b.is_final) {
         // =====================================================================
         // Rules between two non-final impls.
-        //
-        // It's possible to write invalid non-final impls in two files that do
-        // not know about each other, and they are only invalid when they are
-        // both imported and visible together. So we need to compare non-local
-        // impls with each other here.
         // =====================================================================
-        if (DiagnoseNonFinalImplsWithSameTypeStructure(context, impl_a,
-                                                       impl_b)) {
-          continue;
+        if (!did_diagnose_non_final_impls_with_same_type_structure) {
+          if (DiagnoseNonFinalImplsWithSameTypeStructure(context, impl_a,
+                                                         impl_b)) {
+            // The same final `impl_a` may overlap with multiple `impl_b`s, and
+            // we want to diagnose each `impl_b`.
+            did_diagnose_non_final_impls_with_same_type_structure = true;
+          }
         }
       } else if (!impl_a.is_final || !impl_b.is_final) {
-        // No rules that involve a final impl need to be enforced on two
-        // imported impls, so we only diagnose if at least one of the two impls
-        // being compared is local.
-        if (!impl_a.is_local && !impl_b.is_local) {
-          continue;
-        }
-
         // =====================================================================
         // Rules between final impl and non-final impl.
         // =====================================================================
-        if (DiagnoseUnmatchableNonFinalImplWithFinalImpl(
-                context, impl_a_id, impl_a, impl_b_id, impl_b)) {
-          continue;
+        if (!did_diagnose_unmatchable_non_final_impl_with_final_impl) {
+          if (DiagnoseUnmatchableNonFinalImplWithFinalImpl(
+                  context, impl_a_id, impl_a, impl_b_id, impl_b)) {
+            did_diagnose_unmatchable_non_final_impl_with_final_impl = true;
+          }
         }
-      } else if (impl_a.type_structure.IsCompatibleWith(
-                     impl_b.type_structure)) {
-        CARBON_CHECK(impl_a.is_final && impl_b.is_final);
-
-        // No rules that involve a final impl need to be enforced on two
-        // imported impls, so we only diagnose if at least one of the two impls
-        // being compared is local.
-        if (!impl_a.is_local && !impl_b.is_local) {
-          continue;
-        }
-        // =======================================================================
+      } else if (impl_a.type_structure.IsEqualToOrMoreSpecificThan(
+                     impl_b.type_structure) ||
+                 impl_b.type_structure.IsEqualToOrMoreSpecificThan(
+                     impl_a.type_structure)) {
```

Comment:
I believe the previous version of this test was correct and this one is now not. Imagine two final impls, one with a more specific self type and one with a more specific interface. They have overlap, but won't pass this condition. Probably should add a test for this case.

Reply:
Great catch, thank you. Test added and fixed.